clang/lib/Sema/SemaSYCL.cpp

```diff
@@ -2244,34 +2244,38 @@ void SYCLIntegrationHeader::emitFwdDecl(raw_ostream &O, const Decl *D,
     auto *NS = dyn_cast_or_null<NamespaceDecl>(DC);
 
     if (!NS) {
-      if (!DC->isTranslationUnit()) {
-        const TagDecl *TD = isa<ClassTemplateDecl>(D)
-                                ? cast<ClassTemplateDecl>(D)->getTemplatedDecl()
-                                : dyn_cast<TagDecl>(D);
-
-        if (TD && !UnnamedLambdaSupport) {
-          // defined class constituting the kernel name is not globally
-          // accessible - contradicts the spec
-          const bool KernelNameIsMissing = TD->getName().empty();
-          if (KernelNameIsMissing) {
+      const TagDecl *TD = isa<ClassTemplateDecl>(D)
+                              ? cast<ClassTemplateDecl>(D)->getTemplatedDecl()
+                              : dyn_cast<TagDecl>(D);
+      if (!TD)
+        break;
+
+      const bool KernelNameIsMissing = TD->getName().empty();
+      if (KernelNameIsMissing)
+        Diag.Report(KernelLocation, diag::err_sycl_kernel_incorrectly_named)
+            << /* kernel name is missing */ 0;
+      else if (!DC->isTranslationUnit()) {
+        // defined class constituting the kernel name is not globally
+        // accessible - contradicts the spec
+        if (!UnnamedLambdaSupport) {
+          if (TD->isCompleteDefinition())
             Diag.Report(KernelLocation, diag::err_sycl_kernel_incorrectly_named)
-                << /* kernel name is missing */ 0;
-            // Don't emit note if kernel name was completely omitted
-          } else {
-            if (TD->isCompleteDefinition())
-              Diag.Report(KernelLocation,
-                          diag::err_sycl_kernel_incorrectly_named)
-                  << /* kernel name is not globally-visible */ 1;
-            else
-              Diag.Report(KernelLocation, diag::warn_sycl_implicit_decl);
-            Diag.Report(D->getSourceRange().getBegin(),
-                        diag::note_previous_decl)
-                << TD->getName();
-          }
+                << /* kernel name is not globally-visible */ 1;
+          else
+            Diag.Report(KernelLocation, diag::warn_sycl_implicit_decl);
+          Diag.Report(D->getSourceRange().getBegin(), diag::note_previous_decl)
+              << TD->getName();
         }
       }
       break;
     }
+
+    if (NS->isStdNamespace()) {
```

Comment:
This whole thing is a non-issue if unnamed lambda compiler option is enabled even if the user provides a kernel name. Because then we don't template on the lambda name type but on the string representation of it. Therefore we don't need to emit an error in that case.

Reply:
It seems like we could always do that: Template `KernelInfo` on the string representation of the name rather than the name type. Then we would never need to forward declare the name type.

Reply:
I started experimenting with templating KernelInfo on the string representation of the name, and probably went further down that rabbit hole than I should have before I realized I was reimplementing what the existing unnamed lambda logic does already. And so in this latest commit -- meant to prompt discussion -- I flipped a 0 to a 1 (don't laugh :-P) and turned on the unnamed lambda CLI option by default. This seems to deliver exactly what we've been wanting. KernelInfo is templated on the string representation of the name, nullptr_t and other std names work, and the existing code will already skip forward decls when unnamed lambda is on. I assume, though, that the idea of turning on unnamed lambdas by default was already discussed in some meeting somewhere. Is there some reason it can't be this simple?

@rolandschulz @erichkeane @premanandrao @elizabethandrews

Reply:
Right, we shouldn't emit this diagnostic if UnnamedLambdaSupport